h2o-test-support/src/main/java/water/test/WebsocketClient.java

```diff
@@ -0,0 +1,72 @@
+package water.test;
+
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import org.glassfish.tyrus.client.ClientManager;
+import water.H2O;
+
+import javax.websocket.*;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Map;
+
+import static org.junit.Assert.fail;
+
+public class WebsocketClient extends Endpoint {
+
+    private Map<String, String> receivedMessage;
+    private Session sess;
+
+    private final Gson gson = new Gson();
+    
+    public WebsocketClient() throws URISyntaxException, IOException, DeploymentException {
+        String destUri = "ws://" + H2O.getIpPortString() + "/3/Steam.websocket";
+        ClientManager client = ClientManager.createClient();
+        ClientEndpointConfig cec = ClientEndpointConfig.Builder.create().build();
+        sess = client.connectToServer(this, cec, new URI(destUri));
+    }
+    
+    public void close() throws IOException {
+        sess.close(new CloseReason(CloseReason.CloseCodes.NORMAL_CLOSURE, "Test Done"));
+    }
+
+    @OnOpen
+    public void onOpen(Session session, EndpointConfig config) {
+        session.addMessageHandler(String.class, message -> {
+            System.out.println("Received message: " + message);
+            synchronized (this) {
+                receivedMessage = gson.fromJson(message, new TypeToken<Map<String, String>>() {}.getType());
+                this.notifyAll();
+            }
+        });
+    }
+    
+    public void sendMessage(Object msg) throws IOException {
+        sess.getBasicRemote().sendText(gson.toJson(msg));
+    }
+
+    public Map<String, String> waitToReceiveMessage(String message) {
+        return waitToReceiveMessage(message, 10_000);
+    }
+    
+    public Map<String, String> waitToReceiveMessage(String message, int timeout) {
+        return waitToReceiveMessage(message, timeout, true);
+    }
+
+    public synchronized Map<String, String> waitToReceiveMessage(String message, int timeout, boolean failOnNone) {
```

Comment:
If one client can be reused at  multiple places, won't there be a race condition with potential loss of data ? When two threads send a message using the same client (why would the `waitToReceiveMessage` method be synchronized otherwise) at once and then wait for the response, only one response is reacted to.

I bent one of the tests to demonstrate:

```java
    @Test
    public void testSendTwoMessages() throws Exception {
        final WebsocketClient steam = new WebsocketClient();
        try {
            Map<String, String> helloMsg = new HashMap<>();
            helloMsg.put("_type", "hello");
            helloMsg.put("_id", "123456");

            Map<String, String> byeMsg = new HashMap<>();
            byeMsg.put("_type", "bye");
            byeMsg.put("_id", "12345678");

            steam.sendMessage(byeMsg);
            steam.sendMessage(helloMsg);
            System.out.println("Hello message sent.");
            Map<String, String> response = steam.waitToReceiveMessage("hello response");
            Map<String, String> response1 = steam.waitToReceiveMessage("bye response");
            assertEquals("hello_response", response.get("_type"));
            assertEquals("123456_response", response.get("_id"));
            assertEquals(H2O.ABV.projectVersion(), response.get("version"));
            assertEquals(H2O.ABV.branchName(), response.get("branch"));
            assertEquals(H2O.ABV.lastCommitHash(), response.get("hash"));
            assertEquals(String.valueOf(H2O.CLOUD.size()), response.get("cloud_size"));
        } finally {
            steam.close();
        }
    }
```

I'd personally like to have this scenario tested as well (ability of H2O to respond to multiple messages at time).